File: app/src/main/kotlin/com/dilarion/app/ui/screens/auth/AuthViewModel.kt
package com.dilarion.app.ui.screens.auth

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dilarion.app.data.api.ApiService
import com.dilarion.app.data.model.LoginRequest
import com.dilarion.app.monitoring.MonitoringForegroundService
import com.dilarion.app.security.CryptoManager
import com.dilarion.app.security.SessionManager
import com.dilarion.app.services.PresenceService
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import javax.inject.Inject

data class AuthUiState(
    val isLoading: Boolean = false,
    val error: String? = null,
    val success: Boolean = false,
    // Shown once after signup or a recovery-code reset — same reveal screen
    // either way, since both hand back a fresh code that never comes back.
    val revealedUsername: String? = null,
    val revealedRecoveryCode: String? = null,
)

@HiltViewModel
class AuthViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
    private val apiService: ApiService,
    private val sessionManager: SessionManager,
    private val cryptoManager: CryptoManager,
    private val presenceService: PresenceService,
) : ViewModel() {

    private val _uiState = MutableStateFlow(AuthUiState())
    val uiState: StateFlow<AuthUiState> = _uiState

    fun login(username: String, token: String) {
        if (username.isBlank() || token.isBlank()) {
            _uiState.value = AuthUiState(error = "Username and token are required")
            return
        }
        viewModelScope.launch {
            _uiState.value = AuthUiState(isLoading = true)
            runCatching {
                val response = apiService.login(LoginRequest(username.trim(), token.trim()))
                if (!response.isSuccessful) {
                    val msg = when (response.code()) {
                        401  -> "Invalid credentials"
                        500  -> "Server error. Try again."
                        else -> "Login failed (${response.code()})"
                    }
                    _uiState.value = AuthUiState(error = msg)
                    return@launch
                }
                val body = response.body()!!
                val token = body.sessionToken
                if (token.isNullOrBlank()) {
                    _uiState.value = AuthUiState(error = "Server returned no session token")
                    return@launch
                }
                // Generate the E2EE keypair ONCE per device. Regenerating on every
                // login would rotate the identity and make all previously received
                // messages permanently undecryptable.
                val existingPriv = sessionManager.privateKey.first()
                val existingPub = sessionManager.publicKey.first()
                val (pubB64, privB64) = if (existingPriv.isNullOrBlank() || existingPub.isNullOrBlank()) {
                    val kp = cryptoManager.generateKeyPair()
                    apiService.updatePublicKey("Bearer $token", com.dilarion.app.data.model.UpdatePublicKeyRequest(kp.first))
                    kp
                } else {
                    Pair(existingPub, existingPriv)
                }
                sessionManager.saveSession(token, body.username, privB64, pubB64)

                // Register this phone as a device so senders can encrypt to it and
                // we know our device_uuid for finding our entry in a key map.
                runCatching {
                    val resp = apiService.registerDevice(
                        "Bearer $token",
                        com.dilarion.app.data.model.DeviceRegisterRequest(pubB64, "android", android.os.Build.MODEL),
                    )
                    resp.body()?.deviceUuid?.let { sessionManager.saveDeviceUuid(it) }
                }
                presenceService.connect(token)
                MonitoringForegroundService.start(context)
                _uiState.value = AuthUiState(success = true)
            }.onFailure { e ->
                _uiState.value = AuthUiState(error = e.message ?: "Network error")
            }
        }
    }

    fun clearError() { _uiState.value = _uiState.value.copy(error = null) }

    fun clearRevealedCode() {
        _uiState.value = _uiState.value.copy(revealedUsername = null, revealedRecoveryCode = null)
    }

    fun signUp(username: String, phoneNumber: String, token: String) {
        if (username.isBlank() || phoneNumber.isBlank() || token.isBlank()) {
            _uiState.value = AuthUiState(error = "All fields are required")
            return
        }
        viewModelScope.launch {
            _uiState.value = AuthUiState(isLoading = true)
            runCatching {
                val response = apiService.signUp(
                    com.dilarion.app.data.model.SignUpRequest(username.trim(), phoneNumber.trim(), token.trim())
                )
                if (!response.isSuccessful) {
                    val msg = errorBody(response.errorBody()?.string()) ?: "Failed to create account (${response.code()})"
                    _uiState.value = AuthUiState(error = msg)
                    return@launch
                }
                val body = response.body()!!
                _uiState.value = AuthUiState(
                    revealedUsername = body.username,
                    revealedRecoveryCode = body.recoveryCode,
                )
            }.onFailure { e ->
                _uiState.value = AuthUiState(error = e.message ?: "Network error")
            }
        }
    }

    fun resetWithRecoveryCode(username: String, recoveryCode: String, newToken: String) {
        if (username.isBlank() || recoveryCode.isBlank() || newToken.isBlank()) {
            _uiState.value = AuthUiState(error = "All fields are required")
            return
        }
        viewModelScope.launch {
            _uiState.value = AuthUiState(isLoading = true)
            runCatching {
                val response = apiService.resetWithRecoveryCode(
                    com.dilarion.app.data.model.RecoveryCodeResetRequest(username.trim(), recoveryCode.trim(), newToken.trim())
                )
                if (!response.isSuccessful) {
                    val msg = errorBody(response.errorBody()?.string()) ?: "Invalid username or recovery code"
                    _uiState.value = AuthUiState(error = msg)
                    return@launch
                }
                val body = response.body()!!
                _uiState.value = AuthUiState(
                    revealedUsername = body.username,
                    revealedRecoveryCode = body.recoveryCode,
                )
            }.onFailure { e ->
                _uiState.value = AuthUiState(error = e.message ?: "Network error")
            }
        }
    }

    private fun errorBody(raw: String?): String? {
        if (raw.isNullOrBlank()) return null
        return runCatching {
            com.google.gson.JsonParser.parseString(raw).asJsonObject.get("detail")?.asString
        }.getOrNull()
    }
}

File: ios/Dilarion/Sources/Auth/AuthViewModel.swift
import Foundation

struct AuthUiState {
    var isLoading: Bool = false
    var error: String? = nil
    var success: Bool = false
}

class AuthViewModel: ObservableObject {
    @Published var state = AuthUiState()

    // mirrors AuthViewModel.kt: login(username, token)
    func login(username: String, token: String) {
        guard !username.trimmingCharacters(in: .whitespaces).isEmpty,
              !token.trimmingCharacters(in: .whitespaces).isEmpty else {
            state = AuthUiState(error: "Username and token are required")
            return
        }
        state = AuthUiState(isLoading: true)
        Task {
            do {
                let response: LoginResponse = try await APIClient.shared.post(
                    "/auth/login",
                    body: LoginRequest(username: username.trimmingCharacters(in: .whitespaces),
                                       token: token.trimmingCharacters(in: .whitespaces))
                )
                if response.token.isEmpty {
                    await MainActor.run { state = AuthUiState(error: "Server returned no session token") }
                    return
                }
                KeychainHelper.shared.save(key: "session_token", value: response.token)
                KeychainHelper.shared.save(key: "username", value: response.username)

                // Generate the E2EE keypair ONCE per device. Regenerating on every
                // login would rotate the identity and make all previously received
                // messages permanently undecryptable.
                let existingPriv = KeychainHelper.shared.read(key: "private_key")
                let existingPub = KeychainHelper.shared.read(key: "public_key")
                var pubB64: String? = existingPub
                if existingPriv == nil || existingPriv?.isEmpty == true || existingPub == nil || existingPub?.isEmpty == true {
                    if let (pub, priv) = EncryptionManager.shared.generateKeyPair() {
                        KeychainHelper.shared.save(key: "private_key", value: priv)
                        KeychainHelper.shared.save(key: "public_key", value: pub)
                        pubB64 = pub
                        do {
                            struct UpdateKeyRequest: Codable { let public_key: String }
                            let _: EmptyBody = try await APIClient.shared.post(
                                "/users/update_public_key",
                                body: UpdateKeyRequest(public_key: pub)
                            )
                        } catch {
                            print("Failed to upload public key: \(error)")
                        }
                    }
                }

                // Register this phone as a device so senders can encrypt to it and we
                // know our device_uuid for finding our entry in a message key map.
                if let pub = pubB64 {
                    do {
                        let resp: DeviceRegisterResponse = try await APIClient.shared.post(
                            "/devices/register",
                            body: DeviceRegisterRequest(public_key: pub, platform: "ios", device_name: "iPhone")
                        )
                        KeychainHelper.shared.save(key: "device_uuid", value: resp.device_uuid)
                    } catch {
                        print("Device register failed: \(error)")
                    }
                }

                WebSocketManager.shared.connect(token: response.token)
                await MainActor.run { state = AuthUiState(success: true) }
            } catch APIError.serverError(let code, _) {
                let msg: String
                switch code {
                case 401: msg = "Invalid credentials"
                case 500: msg = "Server error. Try again."
                default:  msg = "Login failed (\(code))"
                }
                await MainActor.run { state = AuthUiState(error: msg) }
            } catch {
                await MainActor.run { state = AuthUiState(error: error.localizedDescription) }
            }
        }
    }

    func clearError() { state.error = nil }
}
